### PR TITLE
Add index for storing template models

### PR DESCRIPTION
**Pull request by jmazanec15**
### Description
This PR adds a hidden model index that will be used to stored serialized template indices used during index creation. Some of faiss's index types require a training step before indexing can begin. In order to support these index types, we need a way for a user to train a model index template and serialize it. This template will then be retrieved during segment creation to initialize the faiss index.

The mapping is fairly straightforward:
```
{
  "properties": {
    "engine": {
      "type": "keyword"
    },
    "model_blob": {
      "type": "binary"
    }
  }
}
```
The engine is required to identify which engine the model is intended to be used with. The model_blob stores the binary representation of the model.

1 model maps to 1 OpenSearch document in this index. models are identified by the OpenSearch id field. The index allows users to provide a custom id or allow OpenSearch to generate one for them. Documents cannot be updated in this index, however, they can be removed.

For implementing the index, I referred to:
1. [Anomaly Detection plugin](https://github.com/opensearch-project/anomaly-detection/blob/main/src/main/java/org/opensearch/ad/indices/AnomalyDetectionIndices.java)
2. [Index Management plugin](https://github.com/opensearch-project/index-management/blob/main/src/main/kotlin/org/opensearch/indexmanagement/IndexManagementIndices.kt)

Additionally, I added test cases for each operation.
 
Note -- changes related to jni can be ignored as they are reviewed in #28. The files to be reviewed are:
1. ModelIndex.java
2. KNNSettings.java
3. KNNConstants.java
4. KNNPlugin.java
5. model-mapping.json
6. ModelIndexTests.java 

### Issues Resolved
#27 
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+import static org.opensearch.knn.index.KNNSettings.MODEL_INDEX_NUMBER_OF_REPLICAS_SETTING;
+import static org.opensearch.knn.index.KNNSettings.MODEL_INDEX_NUMBER_OF_SHARDS_SETTING;
+
+/**
+ * ModelIndex is a singleton class that controls operations on the model system index
+ */
+public class ModelIndex {
```
Comment (VijayanB) [inline src/main/java/org/opensearch/knn/indices/ModelIndex.java:51]: final

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    }
+
+    /**
+     * Creates model index. It is possible that the 2 threads call this function simulateously. In this case, one
+     * thread will throw a ResourceAlreadyExistsException. This should be caught and handled.
+     *
+     * @param actionListener CreateIndexResponse listener
+     * @throws IOException thrown when get mapping fails
+     */
+    public void createModelIndex(ActionListener<CreateIndexResponse> actionListener) throws IOException {
```
Comment (VijayanB) [inline src/main/java/org/opensearch/knn/indices/ModelIndex.java:101]: Any reason we don't have to check verify that model doesn't exist before create?

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java (hunk 1 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/indices/ModelIndex.java:51]: Will update

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    }
+
+    /**
+     * Put a model into the system index. Non-blocking. When no id is passed in, OpenSearch will generate the id
+     * automatically. The id can be retrieved in the IndexResponse.
+     *
+     * @param modelBlob byte array of model
+     * @param listener handles index response
+     */
+    public void putModel(KNNEngine knnEngine, byte[] modelBlob, ActionListener<IndexResponse> listener) {
```
Comment (VijayanB) [inline src/main/java/org/opensearch/knn/indices/ModelIndex.java:150]: is it possible to have different name ? method name overloading is not got for readability.

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+     * Get a model from the system index. Call blocks.
+     *
+     * @param modelId to retrieve
+     * @return byte array representing the model
+     * @throws ExecutionException thrown on search
+     * @throws InterruptedException thrown on search
+     */
+    public byte[] getModel(String modelId) throws ExecutionException, InterruptedException {
+        if (!modelIndexExists()) {
+            throw new IllegalStateException("Cannot get model. Model index does not exist.");
```
Comment (VijayanB) [inline src/main/java/org/opensearch/knn/indices/ModelIndex.java:185]: may be add modelID to message?

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java (hunk 2 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/indices/ModelIndex.java:101]: So, the modelIndexExists function can be used before calling this function. There is a chance 2 threads try to create the index. In which case, `ResourceAlreadyExistsException` will be thrown and should be handled. 

I think either approach will work. Do you think this function should include the modelIndexExists check as well? Or should user be responsible for calling modelIndexExists first.

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+                );
+        client.admin().indices().create(request, actionListener);
+    }
+
+    /**
+     * Checks if the model index exists
+     *
+     * @return true if the model index exists; false otherwise
+     */
+    public boolean modelIndexExists() {
```
Comment (VijayanB) [inline src/main/java/org/opensearch/knn/indices/ModelIndex.java:117]: nit: IsModelIndexExist()

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+import static org.opensearch.knn.index.KNNSettings.MODEL_INDEX_NUMBER_OF_SHARDS_SETTING;
+
+/**
+ * ModelIndex is a singleton class that controls operations on the model system index
+ */
+public class ModelIndex {
+    public static Logger logger = LogManager.getLogger(ModelIndex.class);
+
+    //TODO: Should these be stored in constants or in the class? In PR get feedback
+    public static final String MODEL_INDEX_MAPPING = "mappings/model-index.json";
```
Comment (VijayanB) [inline src/main/java/org/opensearch/knn/indices/ModelIndex.java:55]: i prefer constants file

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java (hunk 2 above)
Comment (VijayanB) [reply src/main/java/org/opensearch/knn/indices/ModelIndex.java:101]: nit:Create/Update/Delete instead of CreateModelIndex since the class name is ModelIndex

File: src/test/java/org/opensearch/knn/indices/ModelIndexTests.java [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+        assertFalse(modelIndex.modelIndexExists());
+
+        final CountDownLatch inProgressLatch = new CountDownLatch(1);
+
+        ActionListener<CreateIndexResponse> indexCreationListener = ActionListener.wrap(response -> {
+            assertTrue(response.isAcknowledged());
+            inProgressLatch.countDown();
+        }, exception -> fail("Model index unable to create"));
+
+        modelIndex.createModelIndex(indexCreationListener);
```
Comment (VijayanB) [inline src/test/java/org/opensearch/knn/indices/ModelIndexTests.java:67]: is it possible to create modelIndex manually ? if there is bug in createModelIndex, both create and exists will fail because we are using implementation as part of test set up process.

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java (hunk 2 above)
Comment (VijayanB) [reply src/main/java/org/opensearch/knn/indices/ModelIndex.java:101]: > So, the modelIndexExists function can be used before calling this function. There is a chance 2 threads try to create the index. In which case, `ResourceAlreadyExistsException` will be thrown and should be handled.
> 
> I think either approach will work. Do you think this function should include the modelIndexExists check as well? Or should user be responsible for calling modelIndexExists first.

can this be made singleton to avoid race condition ( check is exists? return false: return true )

Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/indices/ModelIndex.java:101]: So it can be combined into one function, however, it still would not be thread safe. However, this will not cause an issue as long as the exception is caught. Refer to [AD](https://github.com/opensearch-project/anomaly-detection/blob/cfebf07f637c05753e32a1648d05db98e1120e0f/src/main/java/org/opensearch/ad/transport/handler/AnomalyResultBulkIndexHandler.java#L99).

Also, will update naming.

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java (hunk 3 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/indices/ModelIndex.java:150]: I like method name overloading if the operation is the same. I think the parameters differentiate the 2 functions.

I could have `putWithModelId` and `putWithoutModelId` but I think this information is already encompassed in the parameters.

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java (hunk 4 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/indices/ModelIndex.java:185]: Sure

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java (hunk 5 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/indices/ModelIndex.java:117]: isExists doesnt sound right to me, but I have seen that `is` prefix should be used with boolean methods. Therefore, I will change to `isCreated`

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java (hunk 6 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/indices/ModelIndex.java:55]: Will change

File: src/test/java/org/opensearch/knn/indices/ModelIndexTests.java (hunk 7 above)
Comment (jmazanec15) [reply src/test/java/org/opensearch/knn/indices/ModelIndexTests.java:67]: Youre right. It is possible. I will add test functions.

File: src/test/java/org/opensearch/knn/indices/ModelIndexTests.java [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    public void testExists() {
+        ModelIndex modelIndex = ModelIndex.getInstance();
+        assertFalse(modelIndex.isCreated());
+        createIndex(MODEL_INDEX_NAME);
+        assertTrue(modelIndex.isCreated());
+    }
+
+    public void testPut() throws InterruptedException {
```
Comment (VijayanB) [inline src/test/java/org/opensearch/knn/indices/ModelIndexTests.java:70]: nit: shall we split this into multiple test methods? Also, will the created model index is deleted at the end of the test?

Comment (jmazanec15) [reply src/test/java/org/opensearch/knn/indices/ModelIndexTests.java:70]: Yes, I can split into multiple tests

Yes the model index is deleted at the end of each test.

File: src/main/java/org/opensearch/knn/common/KNNConstants.java [hunk 9]
```diff
@@ -51,4 +51,9 @@
     public static final String COMPOUND_EXTENSION = "c";
 
     public static final String JNI_LIBRARY_NAME = "OpensearchKNN";
+
+    public static final String MODEL_BLOB_PARAMETER = "model_blob";
+
+    public static final String MODEL_INDEX_MAPPING_PATH = "mappings/model-index.json";
+    public static final String MODEL_INDEX_NAME = ".knn-model-index";
```
Comment (vamshin) [inline src/main/java/org/opensearch/knn/common/KNNConstants.java:58]: Should the file name start with opensearch? May be confirm other plugins system indices naming convention with openSearch changes.

File: src/main/java/org/opensearch/knn/index/KNNSettings.java [hunk 10]
```diff
@@ -142,6 +144,21 @@
...
             Setting.Property.Deprecated);
 
+    public static final Setting<Integer> MODEL_INDEX_NUMBER_OF_SHARDS_SETTING = Setting.intSetting(
+            MODEL_INDEX_NUMBER_OF_SHARDS,
+            1,
+            1,
+            Setting.Property.NodeScope,
+            Setting.Property.Dynamic);
+
+    public static final Setting<Integer> MODEL_INDEX_NUMBER_OF_REPLICAS_SETTING = Setting.intSetting(
```
Comment (vamshin) [inline src/main/java/org/opensearch/knn/index/KNNSettings.java:154]: Default should be 1 replica?

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+
+    /**
+     * Creates model index. It is possible that the 2 threads call this function simulateously. In this case, one
+     * thread will throw a ResourceAlreadyExistsException. This should be caught and handled.
+     *
+     * @param actionListener CreateIndexResponse listener
+     * @throws IOException thrown when get mapping fails
+     */
+    public void create(ActionListener<CreateIndexResponse> actionListener) throws IOException {
+        if (isCreated()) {
```
Comment (vamshin) [inline src/main/java/org/opensearch/knn/indices/ModelIndex.java:100]: Should we log message here and return?

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+        if (blob == null) {
+            throw new IllegalArgumentException("ModelID: \"" + modelId + "\" is not present in index");
+        }
+
+        return Base64.getDecoder().decode((String) blob);
+    }
+
+    private String getMapping() throws IOException {
+        URL url = ModelIndex.class.getClassLoader().getResource(MODEL_INDEX_MAPPING_PATH);
+        assert url != null;
```
Comment (vamshin) [inline src/main/java/org/opensearch/knn/indices/ModelIndex.java:210]: asserts are sometimes disabled on production hosts. Can we do manual check and throw exception?

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+        GetRequestBuilder getRequestBuilder = new GetRequestBuilder(client, GetAction.INSTANCE, MODEL_INDEX_NAME)
+                .setId(modelId)
+                .setFetchSource(KNNConstants.MODEL_BLOB_PARAMETER, null)
+                .setPreference("_local");
+        GetResponse getResponse = getRequestBuilder.execute().get();
+
+        Object blob = getResponse.getSourceAsMap().get(KNNConstants.MODEL_BLOB_PARAMETER);
+
+        if (blob == null) {
+            throw new IllegalArgumentException("ModelID: \"" + modelId + "\" is not present in index");
```
Comment (vamshin) [inline src/main/java/org/opensearch/knn/indices/ModelIndex.java:202]: Should this message be more user focused? Something like ```There is no model available with the provided Id?```

File: src/main/java/org/opensearch/knn/common/KNNConstants.java (hunk 9 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/common/KNNConstants.java:58]: In [Conventions](https://github.com/opensearch-project/opensearch-plugins/blob/main/CONVENTIONS.md#indices), the only requirement is prefixed with ".".

1. [AD](https://github.com/opensearch-project/anomaly-detection/blob/main/src/main/java/org/opensearch/ad/constant/CommonName.java) is still using "opendistro-"
2. [ISM](https://github.com/opensearch-project/index-management/blob/main/src/main/kotlin/org/opensearch/indexmanagement/IndexManagementPlugin.kt#L189) also prefixes with ".opendistro".

That being said, I think ".opensearch-knn-model-index" is good. I will update.

File: src/main/java/org/opensearch/knn/index/KNNSettings.java (hunk 10 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/KNNSettings.java:154]: Sure, will update.

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java (hunk 11 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/indices/ModelIndex.java:100]: I am a little worried this may blow up the logs if isCreated is not called by caller. Id prefer not to log here.

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java (hunk 12 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/indices/ModelIndex.java:210]: Sure, will update.

File: src/main/java/org/opensearch/knn/indices/ModelIndex.java (hunk 13 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/indices/ModelIndex.java:202]: Sure, I can update.

**Review by vamshin — APPROVED** [review]
LGTM! Thanks
